### PR TITLE
Support returning float and interpolation for aggregate function quantileExactWeighted 

### PR DESCRIPTION
**Pull request by taiyang-li**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Support returning float and interpolation for aggregate function `quantileExactWeighted`.  Since `quantileExactWeighted` had supported interpolation, `quantileInterpolatedWeighted` implementation is not needed any more, besides current interpolation implementation is more accurate than previous `quantileInterpolatedWeighted`.  To keep compatiability for CH users, `quantileInterpolatedWeighted` is treated as alias of `quantileExactWeighted`

Before 
``` sql 
SELECT quantileExactWeighted(0.5)(number, 1)
FROM numbers(4)

Query id: 5a7577b8-3115-4bb3-afc7-dfc402984538

   ┌─quantileExactWeighted(0.5)(number, 1)─┐
1. │                                     1 │
   └───────────────────────────────────────┘
```

After
``` sql
SELECT quantileExactWeighted(0.5)(number, 1)
FROM numbers(4) 

┌─quantileExactWeighted(0.5)(number, 1)─┐
│                                   1.5 │
└───────────────────────────────────────┘ 
```

### CODE REVIEW

**Comment by qoega** [comment]
We need to add aliases for all removed functions `quantileInterpolatedWeighted` etc. to not have breaking changes.

**Comment by rschu1ze** [comment]
I am not too familiar with aggregation functions, @Algunenano may be a better reviewer for this PR.

**Review by Algunenano — CHANGES_REQUESTED** [review]
You can't simply remove an aggregate function, the same way that you can't just change its return type. It's a major compatibility break so it needs better tests and settings:

* Changing the return type is a compatibility break. It should be done carefully and with compatibility settings to allow for seamless migration.
* Please add new test files with multiple tests verifying the feature. Don't add them to old files. Adding new files with new queries, with different types and so on, helps the fuzzer test the code better.
* We need tests to verify that the old states and new states are compatible. For example, this change will break during upgrades if data is inserted with the new version and read with the old one (which is not ok, since rolling upgrades are a must).

File: tests/queries/0_stateless/01663_quantile_weighted_overflow.sql [hunk 1]
```diff
@@ -1,2 +1,4 @@
+-- Tags: no-cpu-aarch64
```
[the remark this answers is not in the document]
Comment (taiyang-li) [reply tests/queries/0_stateless/01663_quantile_weighted_overflow.sql:1]: It is related to undefined behavior, which maybe different on x86 and aarch64. I'll add a comment.

**Comment by taiyang-li** [comment]
@Algunenano  in fact, the former implementation of `quantileInterpolatedWeighted` is not totally accurate if it is compared to other quantile functions. This is the reason why I remove its code and replace `quantileInterpolatedWeighted` with an improved `quantileExactWeighted `.  To keep compatiability, an alias named `quantileInterpolatedWeighted` is added to map to `quantileExactWeighted`

**Comment by taiyang-li** [comment]
@Algunenano hope for your advice, thanks!

**Comment by Algunenano** [comment]
> So interpolation is applied in quantileExactWeighted in this pr, and return type incompatiability is introduced. I think it is unavoidable unless we keep quantileExactWeighted and quantileExactInterpolateWeighted as it was and add a new function like quantileExactInterpolateWeightedNew, which is too ugly to accept.

I don't like introducing a new name either.

No matter what, we need to make sure that the breaking change has a valid upgrade path so that users that have stored states can safely upgrade, and different versions can coexist without issues while the upgrade is taking place. Probably the sanest way to do it is by increasing the version number for this specific aggregate function state and introducing a setting to determine which state is the default (and maybe change the return type based on this setting too).

Then we need to make sure that:
a) new service with setting set to old value is compatible with old services replication
b) new service with setting to new can read old values and transform then into the new values
c) new service can work with mixed old and new values

IMO there is no need to keep the old logic and we can simply cast to the old type if needed, but we can't just break compatibility for something that's stored in disk without a good way to move forward. It's one of the core principles of CH (and the reason why you don't need to dump + restore in every upgrade)

**Comment by taiyang-li** [comment]
@Algunenano let me confirm my new solutions with you 
1. add a settings like `use_interpolation_for_quantile_exact_weighted`, and its default values is false. When it is false, all the behaviors of `quantileExactWeighted` are compatiable with original version.
2. Keep `quantileInterpolatedWeighted` as it was. Maybe we could deprecate it after we have an more accurate interpolation version of `quantileExactWeighted`

**Comment by Algunenano** [comment]
> add a settings like use_interpolation_for_quantile_exact_weighted, and its default values is false. When it is false, all the behaviors of quantileExactWeighted are compatiable with original version.

I would prefer to make it something like `quantileExactWeighted_default_version`, which would be 0 by default, and 1 if you want the new mode. It would be returned by `::getDefaultVersion()`. For a while the server needs to be able to read both version 0 and 1 and to write whatever the default version is (based on the setting). I would suggest making it a server setting, since it doesn't make sense to be able to change it during runtime or per query.

Please make sure you don't inadvertently enable or introduce versioning for other functions (I see it's a convoluted template). Maybe it makes sense to further separate `Weighted` quantiles from normal ones.

> Keep quantileInterpolatedWeighted as it was. Maybe we could deprecate it after we have an more accurate interpolation version of quantileExactWeighted

It's ok for me if it's left unchanged. We can decide in the future whether to deprecate it or improve it via versioning.

**Comment by taiyang-li** [comment]
@Algunenano I had already add new function without changing the behavior of old functions.  https://github.com/ClickHouse/ClickHouse/pull/69619   Please help review if you are convenient. Thanks!

**Pull request closed without merging**